doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,125 @@
+algorithm of the right type, and to be able to process infinite amounts of
+data if the fetched algorithm permits it (typically, algorithms like ED25519
+or RSA-SHA256).
+
+It must be made clear that the added functionality can not be used to
+compose an algorithm from different parts.  For example, it's not possible
+to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
+specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
+`EVP_SIGNATURE` "RSA", the input still be expected to be a digest, or some
+other input that's limited to the modulus size of the RSA pkey.
+
+### For signing with `EVP_SIGNATURE`
+
+New initializers:
+
+``` C
+int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
+                           EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
+```
+
+Added stream functionality:
+
+``` C
+int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
+                        unsigned char *sig, size_t *siglen, size_t sigsize);
+```
+
+### For verifying with `EVP_SIGNATURE`
+
+With verifying, recent development has shown a need to set the signature to
+be verified against separately.  This proposal reflects that.
+
+New initializers:
+
+``` C
+int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
+                             EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
+int EVP_PKEY_verify_recover_init_ex2(EVP_PKEY_CTX *pctx,
+                                     EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
+```
+
+New signature setter:
+
+``` C
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
+                               unsigned char *sig, size_t *siglen, size_t sigsize);
+```
+
+Added stream functionality:
+
+``` C
+int EVP_PKEY_verify_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);
+
+int EVP_PKEY_verify_recover_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
+int EVP_PKEY_verify_recover_final(EVP_PKEY_CTX *ctx);
+```
+
+Requirements on the providers
+-----------------------------
+
+Because it's not immediately obvious from a composite algorithm name what
+key type it requires / supports, at least in code, allowing the use of an
+explicitly fetched implementation of a composite algorithm requires that
+providers cooperate by declaring what key type is required / supported by
+each algorithm.
+
+For non-composite operation algorithms (like "RSA"), this is not necessary,
+see the fallback strategies below.
+
+There are two ways this could be implemented:
+
+1.  through an added provider function that would work like keymgmt's
+    `query_operation_name` function, but would return a key type name
+    instead:
+
+    ``` C
+    # define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
+    OSSL_CORE_MAKE_FUNC(const char *, signature_query_key_type, (void))
+    ```
+
+2.  through a gettable `OSSL_PARAM`, using the param identity "keytype"
```

Comment:
This returns just one type but presumably RSA-PSS-WITH-SHA256 could use both RSA and RSA-PSS keys.

Reply:
Eh gods, that.  How I regret making RSA-PSS a separate key type...

Reply:
I'll have a think about that

Reply:
fyi -- I have been working on at adding RSA-OAEP as a new key type in #22344

Reply:
I don't like it, but I'll reduce this part to have only the first option, and let it return `const char **`.